app/grandchallenge/workstations/static/workstations/js/session-control.js

```diff
@@ -27,20 +29,33 @@ function openWorkstationSession(element) {
             console.warn(err);
         }
 
-        if (workstationWindow === null || isBlankContext) {
+        if (workstationWindow === null || isBlankContext ) {
             window.open(creationURI, windowIdentifier);
         } else {
             workstationWindow.focus();
-            const msg = {
-                sessionControl: {
-                    loadQuery: query
-                }
-            }
-            workstationWindow.postMessage(msg, sessionOrigin);
+
+            const fallback = setTimeout(() => {
+                // Assume window is non-responsive
+                window.open(creationURI, windowIdentifier);
+            }, timeout);
+
+            potentialSessionOrigins.forEach((origin) => {
+                sendSessionControlMessage(workstationWindow, origin, {loadQuery: query}, () => {
```

Comment:
`urlencode` is used at the end of `workstation_query(...)` in `workstations.py` to convert a dictionary of query parameters to a URL string. This is no longer necessary as it is now being communicated over this interface. It would be better to pass the dictionary straight across by removing the final `urlquote` and then handling the changed API in CIRRUS.

This should get over the problem of users having non-unreserved characters in their username which are then quoted (https://github.com/python/cpython/blob/e5b08ddddf1099f04bf65e63017de840bd4b5980/Lib/urllib/parse.py#L835-L838), we saw this example with users who use their email as their username (#2705) but will also occur for non-ascii characters:

```
>>> from urllib.parse import quote_plus
>>> quote_plus("dfasg@")
'dfasg%40'
>>> quote_plus("김")
'%EA%B9%80'
```

Reply:
I've actually considered this before. Mainly as a cleaner way of going about things. However,  we're dependent on it being a query string for the initial session start: we aren't too sure when the viewer can handle the session control so we'd dump it in the URL and let the viewer pick it up from there.

A workaround is to build in a wait for the session to get up and send some ack/ready to the GC JS. But that is a hornet's nest of checks and fallbacks. For instance: in 40 seconds of boot-up time I can get the user to navigate away from the original view spawn page. Alternatively, you can park a first session-control action at the session object level.

Reply:
I see, yes we need to solve it for both the URL bar and message. Lets sit down and see how the components are decoded on the CIRRUS side.